File: app/models/unit.rb
class Unit
  attr_accessor :unit_type, :player, :x, :y,
    :health, :movement_points,
    :attacks, :attacked, :healed, :flank_penalty, :loaded_units,
    :player_id, :summoning_sickness, :moved,
    :build_phase, :current_build

  def initialize(
    unit_type, player, x, y,
    health = 10, movement_points = nil,
    attacks = 0, flank_penalty = 0, loaded_units = [], summoning_sickness = true,
    moved = true, build_phase = nil, current_build = nil, attacked = false,
    healed = false
  )
    @unit_type = unit_type
    @player = player
    @x = x
    @y = y
    @health = health
    @movement_points = movement_points || UnitDefinitions[unit_type][:movement_points]
    @attacks = attacks
    @attacked = attacked
    @healed = healed
    @flank_penalty = flank_penalty
    @loaded_units = loaded_units
    @player_id = nil
    @summoning_sickness = summoning_sickness
    @moved = moved
    @build_phase = build_phase
    @current_build = current_build
  end

  def self.price_for_unit_type(unit_type)
    UnitDefinitions[unit_type][:credits] rescue nil
  end

  def scrap_value
    base = Unit.price_for_unit_type(self.unit_type).to_f / 2.0
    (base * 0.1 * self.health).to_i
  end

  def can_attack_unit_type?(defender_type)
    (self.attack[UnitDefinitions[defender_type][:armor_type]] || 0) > 0
  end

  def has_enough_attack_points_to_attack?
    attacks < attack_phases
  end

  def attack_allowed_by_attack_type?
    attack_type != :exclusive or moved == false
  end

  def calculate_damage(terrain, defender, defender_terrain, defender_capturing = false)
    a = attack[defender.armor_type] + attack_bonus[terrain]
    a = 0 if a < 0 or attack[defender.armor_type] == 0

    d = defender.armor + defender.armor_bonus[defender_terrain]
    d -= 2 if defender_capturing
    d = 0 if d < 0 or defender.armor == 0

    p = 0.05 * (a - d + defender.flank_penalty) + 0.5

    damage = 0
    health.times { damage += 1 if rand < p }

    damage
  end

  def melee_attack?
    self.range[0] == 1
  end

  def can_zoc_unit_type?(defender_type)
    if self.zoc.is_a?(Array)
      self.zoc.include?(UnitDefinitions[defender_type][:armor_type])
    elsif self.zoc
      self.melee_attack? and self.can_attack_unit_type?(defender_type)
    else
      false
    end
  end

  def can_capture?
    self.can_capture
  end

  def loaded_capacity
    self.loaded_units.map do |u|
      self.transport_armor_types[u.armor_type]
    end.sum
  end

  def can_heal_unit_type?(unit_type)
    self.can_heal and self.can_heal.include?(UnitDefinitions[unit_type][:armor_type])
  end

  def copy_attributes_to_unit(unit)
    self.instance_variables.each do |v|
      unit.instance_variable_set(v, self.instance_variable_get(v))
    end
  end

  def tile_index_cost(tile_index)
    UnitDefinitions[@unit_type][:movement][TileTypes[tile_index]]
  end

  def terrain_cost(modified_terrain, actual_terrain)
    if modified_terrain and self.armor_type != :naval
      UnitDefinitions[@unit_type][:movement][modified_terrain]
    else
      UnitDefinitions[@unit_type][:movement][actual_terrain]
    end
  end

  def max_movement_points
    UnitDefinitions[@unit_type][:movement_points]
  end

  def to_json_hash
    h = {}
    self.instance_variables.each do |v|
      h[v.to_s[1..-1].to_sym] = self.instance_variable_get(v)
    end
    h
  end

  def self.from_json_hash(json)
    u = Unit.allocate
    json.each do |k, v|
      u.instance_variable_set(:"@#{k}", v)
    end
    u
  end

  protected

  def method_missing(meth, *args)
    if UnitDefinitions[@unit_type].has_key?(meth)
      UnitDefinitions[@unit_type][meth]
    else
      super
    end
  end

  def respond_to_missing?(meth, *args)
    UnitDefinitions[@unit_type].has_key?(meth)
  end
end

File: public/javascripts/ec/game/unit.js
var UnitLayer = 6;
var OverlayLayer = 8;

var Unit = new JS.Class(CompositeSprite, {
  initialize: function(unit_json) {
    this.callSuper();

    this.def = GameConfig.units[unit_json.unit_type];

    this.unit_type = unit_json.unit_type;
    this.player = unit_json.player;
    this.player_id = unit_json.player_id;
    this.flank_penalty = unit_json.flank_penalty || 0;

    this.loaded_units = [];
    for (var i in unit_json.loaded_units) {
      this.loaded_units.push(new Unit(unit_json.loaded_units[i]));
    }

    this.unit = new Sprite(SpriteSheets.units);
    this.unit.set_sprite(this.unit_type + 'Right', this.player);
    this.unit.set_layer(UnitLayer);
    this.add_sprite(this.unit);

    this.health_indicator = new Sprite(SpriteSheets.health);
    this.health_indicator.set_layer(OverlayLayer);
    this.add_sprite(this.health_indicator);

    this.movement_indicator = new Sprite(SpriteSheets.movement);
    this.movement_indicator.set_layer(OverlayLayer);
    this.add_sprite(this.movement_indicator);

    this.attacks_indicator = new Sprite(SpriteSheets.attacks);
    this.attacks_indicator.set_layer(OverlayLayer);
    this.add_sprite(this.attacks_indicator);

    this.transport_indicator = new Sprite(SpriteSheets.transport_capacity);
    this.transport_indicator.set_layer(OverlayLayer);
    this.add_sprite(this.transport_indicator);

    this.building_indicator = new Sprite(SpriteSheets.unit_building);
    this.building_indicator.set_layer(OverlayLayer);
    this.add_sprite(this.building_indicator);

    if (unit_json.x != undefined && unit_json.y != undefined) this.set_position(unit_json.x, unit_json.y);
    this.set_health(unit_json.health || 10);
    this.set_movement_points(
      unit_json.movement_points == undefined ? this.def.movement_points : unit_json.movement_points
    );
    this.set_attacks(unit_json.attacks || 0);
    this.attacked = unit_json.attacked;
    this.healed = unit_json.healed;
    this.set_transport_capacity();
    this.set_build_phase(unit_json.build_phase);

    this.current_build = unit_json.current_build;

    this.moved = unit_json.moved;
  },

  to_json: function() {
    var j = {
      attacks: this.attacks, flank_penalty: this.flank_penalty, health: this.health,
      loaded_units: [], moved: this.moved, movement_points: this.movement_points,
      player: this.player, player_id: this.player_id, summoning_sickness: this.summoning_sickness,
      unit_type: this.unit_type, x: this.x, y: this.y
    }

    for (var i in this.loaded_units) {
      j.loaded_units.push(this.loaded_units[i].to_json());
    }

    return j;
  },

  add_to_map: function(map) {
    this.callSuper(map, this.x, this.y, this.layer);
    this.set_build_phase(this.build_phase);
  },

  set_position: function(x, y) {
    var old_px = Map.tile_to_pixel_coords((this.x || 0), (this.y || 0))[0];
    var new_px = Map.tile_to_pixel_coords(x, y)[0];
    var facing = (new_px < old_px ? 'Left' : 'Right');

    if (old_px != new_px) this.unit.set_sprite(this.unit_type + facing, this.player);

    this.callSuper(x, y);
  },

  face_unit: function(unit) {
    var tx = Map.tile_to_pixel_coords((this.x || 0), (this.y || 0))[0];
    var ux = Map.tile_to_pixel_coords((unit.x || 0), (unit.y || 0))[0];
    var facing = (ux < tx ? 'Left' : 'Right');

    if (ux != tx) this.unit.set_sprite(this.unit_type + facing, this.player);
  },

  face_right: function() {
    this.unit.set_sprite(this.unit_type + 'Right', this.player);
  },

  face_left: function() {
    this.unit.set_sprite(this.unit_type + 'Left', this.player);
  },

  set_health: function(health) {
    this.health = health;
    this.health_indicator.set_sprite(0, 10 - health);
  },

  set_movement_points: function(mv) {
    this.movement_points = mv;

    var level = 0;
    if (mv == 0) level = 2;
    else if (mv < this.def.movement_points) level = 1;

    this.movement_indicator.set_sprite(0, level);
  },

  set_attacks: function(attacks) {
    this.attacks = attacks;

    var level = 0;
    if (attacks == this.def.attack_phases) level = 2;
    else if (attacks > 0) level = 1;

    this.attacks_indicator.set_sprite(0, level);
  },

  set_transport_capacity: function() {
    var capacity = 0;

    for (var i in this.loaded_units) {
      capacity += this.def.transport_armor_types[this.loaded_units[i].def.armor_type];
    }

    this.transport_indicator.set_sprite(0, capacity);
  },

  set_build_phase: function(phase) {
    this.build_phase = phase;

    if (this.build_phase != null && this.build_phase != undefined && this.build_phase > -1) {
      this.building_indicator.e.show();
    } else {
      this.building_indicator.e.hide();
    }
  },

  scrap_value: function() {
    var base = this.def.credits / 2;
    return Math.floor(base * 0.1 * this.health);
  },

  base_attack_vs_unit: function(unit) {
    var a = this.def.attack[unit.def.armor_type];
    if (a < 0) return 0;
    else return a;
  },

  attack_bonus_from_terrain: function(terrain) {
    return this.def.attack_bonus[terrain];
  },

  base_armor: function() {
    return this.def.armor;
  },

  armor_bonus_from_terrain: function(terrain) {
    return this.def.armor_bonus[terrain];
  },

  attack_vs_unit_from_terrain: function(unit, terrain) {
    var a = this.base_attack_vs_unit(unit);

    if (a > 0) {
      a = a + this.attack_bonus_from_terrain(terrain);
    }

    if (a < 0) return 0;
    else return a;
  },

  armor_from_terrain: function(terrain) {
    return this.base_armor() + this.armor_bonus_from_terrain(terrain);
  },

  base_combat_display: function(defender) {
    return this.base_attack_vs_unit(defender) + '/' + this.base_armor();
  },

  terrain_bonuses_display: function(terrain) {
    var a = this.attack_bonus_from_terrain(terrain);
    var d = this.armor_bonus_from_terrain(terrain);

    return (a >= 0 ? '+' + a : a) + '/' + (d >= 0 ? '+' + d : d);
  },

  flank_penalty_display: function() {
    return '0/-' + this.flank_penalty;
  },

  capture_penalty_display: function() {
    return '0/-2';
  },

  total_combat_display: function(defender, terrain, capturing) {
    var a = this.attack_vs_unit_from_terrain(defender, terrain);
    var d = this.armor_from_terrain(terrain) - this.flank_penalty;

    if (capturing) d -= 2;
    if (d < 0) d = 0;

    return a + '/' + d;
  },

  hit_chance_display: function(defender, terrain, defender_terrain, defender_capturing) {
    var a = this.attack_vs_unit_from_terrain(defender, terrain);
    var d = defender.armor_from_terrain(defender_terrain) - defender.flank_penalty;

    if (defender_capturing) d -= 2;
    if (d < 0) d = 0;

    var base = 0.5;

    var chance = Math.round((base + ((a - d) * 0.05)) * 100);
    if (chance < 0) chance = 0;

    return chance + '%';
  },

  can_attack_unit_type: function(defender_type) {
    return this.def.attack[GameConfig.units[defender_type].armor_type] > 0;
  },

  melee_attack: function() {
    return this.def.range[0] == 1;
  },

  can_zoc_unit_type: function(defender_type) {
    return this.melee_attack() && this.can_attack_unit_type(defender_type);
  },

  can_zoc: function(defender) {
    if (typeof this.def.zoc == 'object') {
      return $.inArray(defender.def.armor_type, this.def.zoc) > -1;
    } else if (this.def.zoc == 'normal') {
      return this.can_zoc_unit_type(defender.unit_type);
    } else {
      return false;
    }
  },

  can_load_unit: function(unit) {
    if (unit.player == this.player && this.def.transport_armor_types) {
      if (this.def.transport_armor_types[unit.def.armor_type]) {
        var capacity = 0;

        for (var i in this.loaded_units) {
          capacity += this.def.transport_armor_types[this.loaded_units[i].def.armor_type];
        }

        if (capacity + this.def.transport_armor_types[unit.def.armor_type] <= this.def.transport_capacity) {
          return true;
        }
      }
    }

    return false;
  },

  can_heal_unit: function(unit) {
    if (unit.player == this.player && this.def.can_heal) {
      if ($.inArray(unit.def.armor_type, this.def.can_heal) > -1) {
        if (unit.health < 10) {
          return true;
        }
      }
    }

    return false;
  },

  get_x: function() {
    return this.x;
  },

  get_y: function() {
    return this.y;
  },

  tile_index_cost: function(tile_index) {
    return this.def.movement[GameConfig.tiles[tile_index]];
  },

  terrain_cost: function(modified_terrain, actual_terrain) {
    if (modified_terrain && this.def.armor_type != 'naval') {
      return this.def.movement[modified_terrain];
    } else {
      return this.def.movement[actual_terrain];
    }
  },

  get_movement_points: function() {
    return this.movement_points;
  },

  max_movement_points: function() {
    return this.def.movement_points;
  },

  get_max_range: function() {
    return this.def.range[1];
  },

  get_min_range: function() {
    return this.def.range[0];
  },
});
